File: src_old/flet_pages/feedback_page.py
import flet as ft
import system_data
import firestore_db
import time

class FeedbackPage(ft.View):
    def __init__(self, page: ft.Page, pages_visited) -> None:
        super().__init__()
        # CONSTRUCT THE PAGE
        # Assign passed globals to instance
        self.page               = page
        self.page.theme_mode    = ft.ThemeMode.DARK
        self.pages_visited      = pages_visited
        self.scroll             = ft.ScrollMode.ALWAYS
        ############################################################
        # Define General Data

        ############################################################
        # Define Icons
        self.menu_icon      = ft.Icon(
            name    = ft.Icons.MENU_SHARP, 
            color   = ft.Colors.BLACK)
        self.buy_me_a_coffee_image = ft.Image(src="system_data/non_question_assets/support_me.webp",
                                              width = 300)        
        ############################################################
        # Define Simple UI components
        self.message_to_user = ft.Text(value="Thanks for Downloading Quizzer!\n\n I assume you're here to give feedback, maybe a bug, or a suggestion for a new feature. I look forward to hearing back. So long as you have an internet connection I'll be able to read whatever it is you submit. And if you're really enjoying the experience or would just like to help fund further development, I'd be grateful if you would donate using the Buy Me A Coffee link down below.\n\n The aim of quizzer is to simplify and enhance the learning process, so certain features like building 'decks' or similar things that support cramming are just not intended functionality. Learning is lifelong endevour and quizzer is your companion on that endevour. So while I understand that you might be looking to just cram for that next test I'd highly encourage you to adopt a system that will help you retain what you've learned over the long term, not just long enough to pass that exam.",
                                       size = 16)
        self.categorize_menu    = ft.Dropdown(
            label       = "Type of Feedback",
            hint_text   = "Is this a Bug Report, A Feature, or something else?",
            options     = [
                ft.dropdown.Option("Bug Report"),
                ft.dropdown.Option("Feature Suggestion"),
                ft.dropdown.Option("Something Else")
            ]
        )
        self.feedback_box       = ft.TextField(
            label       = "What are you thinking?",
            multiline   = True
        )
        self.donation_link      = ft.Container(
            content=self.buy_me_a_coffee_image,
            ink             = True,
            ink_color       = ft.Colors.GREY_500,
            bgcolor         = ft.Colors.GREEN,
            border_radius   = 50,
            on_click        = self.open_link
        )
        # Components with functions
        self.menu_button                    = ft.ElevatedButton(
            content=self.menu_icon, 
            bgcolor="white", 
            on_click=self.go_to_menu_page)
        self.exit_button                    = ft.IconButton(
            icon=ft.Icons.ARROW_BACK,
            icon_color=ft.Colors.BLACK,
            bgcolor=ft.Colors.WHITE,
            on_click=self.go_to_home_screen
        )
        self.submit_button                  = ft.ElevatedButton(
            text        = "Submit Feedback",
            on_click    = self.submit_feedback
        ) 
        ############################################################
        # Define Composite Components -> ft.Container buttons

        ############################################################
        # Define Container elements for organization
        self.header_row = ft.Row(
            alignment=ft.MainAxisAlignment.SPACE_BETWEEN,
            controls=[
                self.menu_button,
                ft.Text(value="Give Feedback!", size=25),
                self.exit_button
            ]
        )
        ############################################################
        # Piece it all together with self.controls
        self.controls=[
            self.header_row,
            self.message_to_user,
            self.categorize_menu,
            self.feedback_box,
            self.submit_button,
            ft.Row(alignment=ft.MainAxisAlignment.CENTER, controls=[self.donation_link])
        ]
    # Page Functionality below:
    # Navigation Functions Built In
    def open_link(self, e = None):
        self.page.launch_url("https://buymeacoffee.com/quizer")

    def submit_feedback(self, e: ft.ControlEvent):
        print(self.categorize_menu.value)
        print(self.feedback_box.value)
        self.submit_button.text     = "THANKS!"
        self.submit_button.disabled = True
        self.page.update()
        time.sleep(1)
        self.submit_button.disabled = False
        category = self.categorize_menu.value
        feedback = self.feedback_box.value
        firestore_db.submit_feedback_to_firestore(category, feedback)
        self.categorize_menu.value = "Is this a Bug Report, A Feature, or something else?"
        self.feedback_box.value    = "Anything Else?"
        self.page.update()
    def go_to_new_profile_screen            (self, e: ft.ControlEvent = None):
        self.page.go("/NewProfilePage")

    def go_to_home_screen                   (self, e: ft.ControlEvent = None):
        self.page.go("/HomePage")

    def go_to_login_page                    (self, e: ft.ControlEvent = None):
        self.page.go("/LoginPage")

    def go_to_menu_page                     (self, e: ft.ControlEvent = None):
        self.page.go("/Menu")

    def go_to_add_question_page             (self, e: ft.ControlEvent = None):
        self.page.go("/AddQuestionPage")

    def go_to_edit_question_page            (self, e: ft.ControlEvent = None):
        self.page.go("/EditQuestionPage")

    def go_to_settings_page                 (self, e: ft.ControlEvent = None):
        self.page.go("/SettingPage")

    def go_to_stats_page                    (self, e: ft.ControlEvent = None):
        self.page.go("/StatsPage")

    def go_to_user_profile_page             (self, e: ft.ControlEvent = None):
        self.page.go("/UserProfilePage")

    def go_to_display_modules_page             (self, e: ft.ControlEvent = None):
        self.page.go("/DisplayModulePage")
    
    def go_to_ai_question_generator_page    (self, e: ft.ControlEvent = None):
        self.page.go("/AIQuestionGeneratorPage")

File: quizzer/lib/UI_systems/09_feedback_page/feedback_page.dart
import 'package:flutter/material.dart';
import 'package:quizzer/UI_systems/global_widgets/widget_global_app_bar.dart';
import 'package:quizzer/backend_systems/logger/quizzer_logging.dart';
import 'package:url_launcher/url_launcher.dart';
import 'dart:async'; // For Timer
import 'package:quizzer/backend_systems/session_manager/session_manager.dart';
import 'package:quizzer/app_theme.dart';

class FeedbackPage extends StatefulWidget {
  const FeedbackPage({super.key});

  @override
  State<FeedbackPage> createState() => _FeedbackPageState();
}

class _FeedbackPageState extends State<FeedbackPage> {
  String? _selectedCategory;
  final TextEditingController _feedbackController = TextEditingController();
  bool _isSubmitting = false;
  String _submitButtonText = "Submit Feedback";

  final List<String> _feedbackCategories = [
    "Bug Report",
    "Feature Suggestion",
    "Something Else"
  ];

  Future<void> _launchBuyMeACoffee() async {
    final Uri url = Uri.parse('https://buymeacoffee.com/quizer');
    if (!await launchUrl(url)) {
      QuizzerLogger.logError('Could not launch $url');
      if (mounted) {
        ScaffoldMessenger.of(context).showSnackBar(
          const SnackBar(content: Text('Could not open link. Please try again later.')),
        );
      }
    }
  }

  void _submitFeedback() async {
    if (_selectedCategory == null) {
      ScaffoldMessenger.of(context).showSnackBar(
        const SnackBar(content: Text('Please select a category.')),
      );
      return;
    }
    if (_feedbackController.text.trim().isEmpty) {
      ScaffoldMessenger.of(context).showSnackBar(
        const SnackBar(content: Text('Please enter your feedback.')),
      );
      return;
    }

    setState(() {
      _isSubmitting = true;
      _submitButtonText = "THANKS!";
    });

    try {
      QuizzerLogger.logMessage(
          "Sending feedback via SessionManager: Category: $_selectedCategory, Feedback: ${_feedbackController.text}");
      SessionManager().submitUserFeedback(
        feedbackType: _selectedCategory!, // Assert non-null as it passed validation
        feedbackContent: _feedbackController.text.trim(),
      ); // The Future is intentionally not awaited here to avoid blocking UI
      QuizzerLogger.logSuccess("Feedback submission initiated via SessionManager.");

      // Reset form and show success message on successful submission
      if (mounted) {
        setState(() {
          _isSubmitting = false;
          _submitButtonText = "Submit Feedback";
          _selectedCategory = null;
          _feedbackController.clear();
        });
        ScaffoldMessenger.of(context).showSnackBar(
          const SnackBar(content: Text('Feedback submitted successfully!')),
        );
      }
    } catch (e, stackTrace) {
      QuizzerLogger.logError(
          'Error submitting feedback via SessionManager: $e\nStack Trace: $stackTrace' 
      );
      if (mounted) {
        setState(() {
          _isSubmitting = false; // Reset submitting state on error
          _submitButtonText = "Submit Feedback";
        });
        ScaffoldMessenger.of(context).showSnackBar(
          SnackBar(content: Text('Error submitting feedback: ${e.toString()}')),
        );
      }
    } finally {
      // The Timer for UI reset is removed as setState for success/error handles it.
    }
  }

  @override
  void dispose() {
    _feedbackController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: const GlobalAppBar(
        title: 'Give Feedback!',
        showHomeButton: true,
      ),
      body: SingleChildScrollView(
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: <Widget>[
            const Text(
              "Thanks for Downloading Quizzer!\n\nI assume you're here to give feedback, maybe a bug, or a suggestion for a new feature. I look forward to hearing back. So long as you have an internet connection I'll be able to read whatever it is you submit. And if you're really enjoying the experience or would just like to help fund further development, I'd be grateful if you would donate using the Buy Me A Coffee link down below.\n\nThe aim of quizzer is to simplify and enhance the learning process, so certain features like building 'decks' or similar things that support cramming are just not intended functionality. Learning is lifelong endevour and quizzer is your companion on that endevour. So while I understand that you might be looking to just cram for that next test I'd highly encourage you to adopt a system that will help you retain what you've learned over the long term, not just long enough to pass that exam.\n\n Donations help keep quizzer free for users who can't afford it otherwise",
            ),
            AppTheme.sizedBoxLrg,
            const Text('Type of Feedback'),
            DropdownButton<String>(
              value: _selectedCategory,
              hint: const Text('Is this a Bug Report, A Feature, or something else?'),
              isExpanded: true,
              underline: const SizedBox.shrink(),
              items: _feedbackCategories.map((String category) {
                return DropdownMenuItem<String>(
                  value: category,
                  child: Text(category),
                );
              }).toList(),
              onChanged: (String? newValue) {
                setState(() {
                  _selectedCategory = newValue;
                });
              },
            ),
            AppTheme.sizedBoxMed,
            TextField(
              controller: _feedbackController,
              style: TextStyle(color: Theme.of(context).colorScheme.onPrimary),
              decoration: const InputDecoration(
                labelText: 'What are you thinking?',
                hintText: 'Enter your feedback here...',
              ),
              maxLines: 5,
            ),
            AppTheme.sizedBoxLrg,
            Center(
              child: ElevatedButton(
                onPressed: _isSubmitting ? null : _submitFeedback,
                child: Text(_submitButtonText),
              ),
            ),
            AppTheme.sizedBoxLrg,
            Center(
              child: InkWell(
                onTap: _launchBuyMeACoffee,
                child: Image.asset(
                  'images/quizzer_assets/support_me.webp',
                  errorBuilder: (context, error, stackTrace) {
                    QuizzerLogger.logError('Failed to load support_me.webp: $error');
                    return const Text('Could not load image');
                  },
                ),
              ),
            ),
            AppTheme.sizedBoxLrg,
          ],
        ),
      ),
    );
  }
} 